iSCSI HA storage cluster/SAN: drbd, pacemaker, corosync, iSCSI LIO

ryantm opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

drbd used to contain the kernel module code and the userspace
tools. The kernel module continues to be available in the drbd
project, but is also included with the linux kernel directly, so is
probably unnecessary to include. The userpsace tools were split into
drbd-utils, which are packaged here.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jslight90 committed:
nixos/drbd: fix syntax for etc config file

jslight90 committed:
nixos/drbd: fix configuration of systemd service

ryantm committed:
pacemaker: init at 2.0.5

ryantm committed:
resource-agents: init at 4.8.0

ryantm committed:
drbd: remove; drbd-utils: init at 9.16.0

drbd used to contain the kernel module code and the userspace
tools. The kernel module continues to be available in the drbd
project, but is also included with the linux kernel directly, so is
probably unnecessary to include. The userpsace tools were split into
drbd-utils, which are packaged here.

ryantm commented:
Result of `nixpkgs-review pr 119904` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>libvmi</li>
    <li>qemu_xen (qemu_xen_4_10)</li>
    <li>qemu_xen-light (qemu_xen_4_10-light)</li>
    <li>qubes-core-vchan-xen</li>
    <li>xen</li>
    <li>xen-light</li>
    <li>xen-slim</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>drbd (drbd-utils)</li>
    <li>resource-agents</li>
  </ul>
</details>

aanderse commented:
ping @jslight90

SuperSandro2000 commented on pkgs/os-specific/linux/drbd-utils/default.nix:
```diff
@@ -0,0 +1,122 @@
+{ lib, stdenv
```
```suggestion
{ lib
, stdenv
```

SuperSandro2000 commented on pkgs/os-specific/linux/drbd-utils/default.nix:
```diff
@@ -0,0 +1,122 @@
+    sha256 = "sha256:0s1dgg4nvvvxzh9qrah3n15b92h38b8m9pa43s14fdjdg11m6wk2";
+  };
+
+  # patches = [ ./pass-force.patch ./fix-glibc-compilation.patch ];
```
Whats with this?

ryantm replied:
Those are old patches from the drbd derivation. Removed it.

SuperSandro2000 commented on pkgs/os-specific/linux/drbd-utils/default.nix:
```diff
@@ -0,0 +1,122 @@
+  ];
+
+  makeFlags = [
+    "SHELL=${stdenv.shell}"
```
Should we use runtimeShell here?

ryantm replied:
Looks like it wasn't necessary, so I removed it.

SuperSandro2000 commented on pkgs/os-specific/linux/resource-agents/default.nix:
```diff
@@ -0,0 +1,37 @@
+{ lib, stdenv
```
```suggestion
{ lib
, stdenv
```

jslight90 commented:
> ping @jslight90

I'm working on setting up a SAN system with these changes.  I'll update with results within the next couple weeks.

ryantm committed:
drbd-utils: fixes

aanderse commented on nixos/modules/services/network-filesystems/drbd.nix:
```diff
@@ -47,18 +47,33 @@ let cfg = config.services.drbd; in
+      };
+      preStart = ''
+        # Check the configuration file for syntax errors
+        ${pkgs.drbd-utils}/bin/drbdadm sh-nop
```
Would `drbdadm` allow us to run this inside the sandbox and present build time errors?

ryantm replied:
Maybe! But I'd prefer to not delay this to figure that out.

mweinelt commented:
Are the drbd-utils 9.x compatible with the 8.4.x kmods that are intree? AFAIK recent 9.x versions have not been upstreamed into the Linux kernel.

ryantm commented:
> Are the drbd-utils 9.x compatible with the 8.4.x kmods that are intree? AFAIK recent 9.x versions have not been upstreamed into the Linux kernel.

@mweinelt The drbd-utils 9.x [compilation options](https://github.com/LINBIT/drbd-utils/blob/2b895682d20cb3f804f4ec4b30297f67cec3827a/configure.ac#L62) to disable support for 8.3 and 8.4 so I assume that means it is backward compatible. I am in the process of testing it still though.

ryantm committed:
nixos/drbd: cleanup module; make drbd-utils available on service path

Use `adjust all` instead of `up all` because `up all` can fail if
something is already up.

ryantm commented:
It seems to be backward compatible in that it calls out to the correct version binary:
```
# drbdadm -v adjust all
drbdsetup-84 new-resource r0
drbdsetup-84 new-minor r0 0 0
drbdmeta 0 v08 /dev/md127 internal apply-al
drbdsetup-84 attach 0 /dev/md127 /dev/md127 internal
drbdsetup-84 connect r0 ipv4:192.168.1.2:7789 ipv4:192.168.1.2:7789 --protocol=C
```

SuperSandro2000 commented on pkgs/os-specific/linux/drbd-utils/default.nix:
```diff
@@ -0,0 +1,120 @@
+  postPatch = ''
+    patchShebangs .
+    substituteInPlace user/v84/drbdadm_usage_cnt.c \
+      --replace '			"/lib/drbd");' \
+                '			"${placeholder "out"}/lib/drbd");'
+    substituteInPlace user/v9/drbdsetup_linux.c \
+      --replace '		ret = system("/sbin/modprobe drbd");' \
+                '		ret = system("${kmod}/bin/modprobe drbd");'
```
Please dont mix tabs and spaces.

SuperSandro2000 commented on pkgs/os-specific/linux/drbd-utils/default.nix:
```diff
@@ -0,0 +1,120 @@
+    substituteInPlace documentation/ra2refentry.xsl \
+      --replace "http://www.oasis-open.org/docbook/xml/4.4/docbookx.dtd" \
+                "${docbook_xml_dtd_44}/xml/dtd/docbook/docbookx.dtd"
+    substituteInPlace documentation/v9/drbd.conf.xml.in \
+      --replace "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd" \
+                "${docbook_xml_dtd_45}/xml/dtd/docbook/docbookx.dtd"
+    substituteInPlace documentation/v9/drbdsetup.xml.in \
+      --replace "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd" \
+                "${docbook_xml_dtd_45}/xml/dtd/docbook/docbookx.dtd"
+    substituteInPlace documentation/v84/drbdsetup.xml \
+      --replace "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd" \
+                "${docbook_xml_dtd_45}/xml/dtd/docbook/docbookx.dtd"
+    substituteInPlace documentation/v84/drbd.conf.xml \
+      --replace "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd" \
+                "${docbook_xml_dtd_45}/xml/dtd/docbook/docbookx.dtd"
```
This could be simplified with for f in; do .. done.

jslight90 commented:
I was able to setup a highly-available SAN using these drbd changes, pacemaker, corosync, and iSCSI.  I plan to submit a pull request with the pacemaker pkg, and a module for a HA SAN (depends on these changes).

jslight90 commented:
One problem I ran into, the OCF files to manage resources were not in the standard OCF directory (`/usr/lib/ocf`).  I'm not sure whether copying those files into place should be a part of this PR or the HA SAN module.

ryantm committed:
ocf-resource-agents: init

jslight90 commented on pkgs/misc/logging/pacemaker/default.nix:
```diff
@@ -0,0 +1,89 @@
+, pkg-config
+, python3
+
+# Pacemaker is compiled twice, once with forOCR = true to extract its
+# OCF definitions for use in the ocf-resource-agents derivation, then
+# again with forOCR = false, where the ocf-resource-agents is provided
```
Typos:
```suggestion
# Pacemaker is compiled twice, once with forOCF = true to extract its
# OCF definitions for use in the ocf-resource-agents derivation, then
# again with forOCF = false, where the ocf-resource-agents is provided
```

ryantm replied:
thanks; fixed.

jslight90 commented on pkgs/os-specific/linux/drbd-utils/default.nix:
```diff
@@ -0,0 +1,121 @@
+, docbook_xml_dtd_45
+, kmod
+
+# drbd-utils are compiled twice, once with forOCR = true to extract
+# its OCF definitions for use in the ocf-resource-agents derivation,
+# then again with forOCR = false, where the ocf-resource-agents is
```
Typos:
```suggestion
# drbd-utils are compiled twice, once with forOCF = true to extract
# its OCF definitions for use in the ocf-resource-agents derivation,
# then again with forOCF = false, where the ocf-resource-agents is
```

ryantm commented:
> One problem I ran into, the OCF files

I've resolved the issue with OCF files by building drbd and pacemaker twice. Once to extract the OCF files and once with all the combined OCF files from a derivation called `ocf-resource-agents`. It is no longer necessary to put files at `/usr/lib/ocf`.

ryantm committed:
pacemaker: fix non-linux evaluation

ryantm commented:
I changed the title because we are going to use this PR to update all the components and modules we are using for our HA storage cluster/SAN.

ryantm committed:
drbd: add nixos test

astro commented:
I was able to build a working Pacemaker cluster with the packages in this closed PR.

@ryantm @jslight90 What are your future plans? Is this up for grabs?

ryantm commented:
@astro Feel free to pick it up and use what we have here! Currently, I'm evaluating using Proxmox instead, since I was able to get going in almost no time.

astro commented:
Interesting. My motivation is replacing a Frankenproxmox installation with something well-defined.

ryantm commented:
@astro Interesting, we should talk more! I messaged you on Matrix.
